File: go/sign_text_test.go
//go:build cgo

// Per-language sign_text / verify_text tests for the Go SDK (Issue 002).
//
// Mirrors `python/tests/test_sign_text.py` and
// `node/tests/sign-text.test.ts`. Exercises the full Go SDK code path.
//
// CGo build only.

package haiai

import (
	"context"
	"os"
	"path/filepath"
	"strings"
	"testing"
)

func TestSignTextRoundTrip(t *testing.T) {
	cl := mediaParityClient(t)
	ctx := context.Background()

	dir := t.TempDir()
	path := filepath.Join(dir, "hello.md")
	if err := os.WriteFile(path, []byte("# Hello\n"), 0o644); err != nil {
		t.Fatalf("write md: %v", err)
	}

	outcome, err := cl.SignText(ctx, path, SignTextOptions{})
	if err != nil {
		t.Fatalf("SignText: %v", err)
	}
	if outcome.SignersAdded != 1 {
		t.Fatalf("expected signers_added=1, got %d", outcome.SignersAdded)
	}

	verified, err := cl.VerifyText(ctx, path, VerifyTextOptions{})
	if err != nil {
		t.Fatalf("VerifyText: %v", err)
	}
	if verified.Status != "signed" {
		t.Fatalf("expected status=signed, got %s", verified.Status)
	}
	if len(verified.Signatures) != 1 {
		t.Fatalf("expected 1 signature, got %d", len(verified.Signatures))
	}
	if verified.Signatures[0].Status != "valid" {
		t.Fatalf("expected signature status=valid, got %s", verified.Signatures[0].Status)
	}
}

func TestSignTextDefaultBackupCreatesBakFile(t *testing.T) {
	// Issue 003 regression: default Go SignTextOptions{} (zero value) must
	// produce backup=true at the wire layer. Previously, marshalling the
	// struct produced backup=false, silently disabling backup writes.
	cl := mediaParityClient(t)
	ctx := context.Background()

	dir := t.TempDir()
	path := filepath.Join(dir, "with-bak.md")
	if err := os.WriteFile(path, []byte("# original\n"), 0o644); err != nil {
		t.Fatalf("write md: %v", err)
	}

	if _, err := cl.SignText(ctx, path, SignTextOptions{}); err != nil {
		t.Fatalf("SignText: %v", err)
	}
	if _, err := os.Stat(path + ".bak"); err != nil {
		t.Fatalf("expected %s.bak after default SignText, got: %v", path, err)
	}
}

func TestSignTextNoBackupSkipsBak(t *testing.T) {
	// Issue 003: NoBackup=true must propagate to backup=false on the wire.
	cl := mediaParityClient(t)
	ctx := context.Background()

	dir := t.TempDir()
	path := filepath.Join(dir, "no-bak.md")
	if err := os.WriteFile(path, []byte("# no backup\n"), 0o644); err != nil {
		t.Fatalf("write md: %v", err)
	}

	if _, err := cl.SignText(ctx, path, SignTextOptions{NoBackup: true}); err != nil {
		t.Fatalf("SignText with NoBackup: %v", err)
	}
	if _, err := os.Stat(path + ".bak"); err == nil {
		t.Fatalf("expected no .bak file when NoBackup=true, but %s.bak exists", path)
	} else if !strings.Contains(err.Error(), "no such file") {
		t.Fatalf("unexpected stat error: %v", err)
	}
}

func TestVerifyTextStrictMissingSignatureReturnsError(t *testing.T) {
	cl := mediaParityClient(t)
	ctx := context.Background()

	dir := t.TempDir()
	path := filepath.Join(dir, "unsigned.md")
	if err := os.WriteFile(path, []byte("# untouched\n"), 0o644); err != nil {
		t.Fatalf("write md: %v", err)
	}

	// Strict mode raises through the FFI when no signature block is found.
	if _, err := cl.VerifyText(ctx, path, VerifyTextOptions{Strict: true}); err == nil {
		t.Fatalf("expected error from VerifyText(strict=true) on unsigned file")
	}
}

File: node/tests/sign-text.test.ts
/**
 * Per-language sign_text / verify_text tests for the Node SDK (Issue 002).
 *
 * Mirrors `python/tests/test_sign_text.py`. Exercises the full Node SDK path
 * end-to-end via haiinpm native FFI -> binding-core -> JACS.
 *
 * Skipped when haiinpm is missing or the native binding predates TASK_008.
 */

import { afterAll, beforeAll, describe, it, expect } from 'vitest';
import { createRequire } from 'node:module';
import {
  stageSignedMediaVerifier,
  VERIFIER_AGENT_PASSWORD,
} from './signed-verifier-fixture.js';
import {
  mkdtempSync,
  realpathSync,
  writeFileSync,
  rmSync,
  existsSync,
} from 'node:fs';
import { dirname, join, resolve } from 'node:path';
import { fileURLToPath } from 'node:url';
import { tmpdir } from 'node:os';

const __filenameLocal = fileURLToPath(import.meta.url);
const __dirnameLocal = dirname(__filenameLocal);

const REPO_ROOT = resolve(__dirnameLocal, '../..');

process.env.JACS_PRIVATE_KEY_PASSWORD = VERIFIER_AGENT_PASSWORD;

interface NativeHaiClientCtor {
  new (configJson: string): {
    signText?: (path: string, optsJson: string) => Promise<string>;
    verifyText?: (path: string, optsJson: string) => Promise<string>;
  };
}

interface HaiinpmModule {
  HaiClient: NativeHaiClientCtor;
}

function loadHaiinpm(): HaiinpmModule | null {
  try {
    const dynamicRequire = createRequire(__filenameLocal);
    return dynamicRequire('haiinpm') as HaiinpmModule;
  } catch {
    return null;
  }
}

function probeMediaSupport(): boolean {
  const m = loadHaiinpm();
  if (!m) return false;
  try {
    const c = new m.HaiClient('{"jacs_id":"x"}');
    return typeof c.signText === 'function' && typeof c.verifyText === 'function';
  } catch {
    return false;
  }
}

const MEDIA_SUPPORTED = probeMediaSupport();
const SKIP_REASON =
  'Installed haiinpm native binding does not expose the Layer-8 media methods. ' +
  'Rebuild via `cargo build -p haiinpm --release`.';

interface FFIClient {
  signText: (path: string, optsJson: string) => Promise<string>;
  verifyText: (path: string, optsJson: string) => Promise<string>;
}

function buildFFIClient(): { client: FFIClient; tmpDir: string } {
  const haiinpm = loadHaiinpm();
  if (!haiinpm) throw new Error('haiinpm not loadable');
  process.env.JACS_PRIVATE_KEY_PASSWORD = VERIFIER_AGENT_PASSWORD;

  const staged = stageSignedMediaVerifier(REPO_ROOT, 'haiai-sign-text-');
  const ffiConfig = JSON.stringify({
    jacs_id: staged.jacsId,
    agent_name: 'MediaFixtureVerifier',
    agent_version: '1.0.0',
    key_dir: staged.keyDir,
    jacs_config_path: staged.configPath,
    base_url: 'http://localhost:1',
  });
  const native = new haiinpm.HaiClient(ffiConfig);
  return {
    client: native as unknown as FFIClient,
    tmpDir: staged.tmpDir,
  };
}

let ffiClient: FFIClient | null = null;
let agentTmpDir: string | null = null;
let stageTmpDir: string | null = null;

const describeMaybe = MEDIA_SUPPORTED ? describe : describe.skip;

describeMaybe('Node SDK signing-side parity (sign_text)', () => {
  beforeAll(() => {
    if (!MEDIA_SUPPORTED) return;
    const built = buildFFIClient();
    ffiClient = built.client;
    agentTmpDir = built.tmpDir;
    stageTmpDir = realpathSync(mkdtempSync(join(tmpdir(), 'haiai-sign-text-stage-')));
  });

  afterAll(() => {
    if (agentTmpDir) rmSync(agentTmpDir, { recursive: true, force: true });
    if (stageTmpDir) rmSync(stageTmpDir, { recursive: true, force: true });
  });

  if (!MEDIA_SUPPORTED) {
    it.skip(SKIP_REASON, () => {});
    return;
  }

  it('sign_text round-trip: sign + verify a markdown file', async () => {
    const path = join(stageTmpDir!, 'hello.md');
    writeFileSync(path, '# Hello\n');

    const outcomeJson = JSON.parse(await ffiClient!.signText(path, '{}'));
    expect(outcomeJson.signers_added).toBe(1);

    const verifyJson = JSON.parse(await ffiClient!.verifyText(path, '{}'));
    expect(verifyJson.status).toBe('signed');
    expect(verifyJson.signatures).toHaveLength(1);
    expect((verifyJson.signatures as Array<Record<string, unknown>>)[0].status).toBe('valid');
  });

  it('sign_text default backup creates .bak file', async () => {
    const path = join(stageTmpDir!, 'with-bak.md');
    writeFileSync(path, '# original\n');

    await ffiClient!.signText(path, '{}');
    expect(existsSync(`${path}.bak`)).toBe(true);
  });

  it('sign_text with backup=false skips .bak file', async () => {
    const path = join(stageTmpDir!, 'no-bak.md');
    writeFileSync(path, '# no backup\n');

    await ffiClient!.signText(path, JSON.stringify({ backup: false }));
    expect(existsSync(`${path}.bak`)).toBe(false);
  });

  it('verify_text strict mode on missing signature signals failure', async () => {
    const path = join(stageTmpDir!, 'unsigned.md');
    writeFileSync(path, '# untouched\n');

    // Strict mode raises through the FFI as a JACS error.
    await expect(
      ffiClient!.verifyText(path, JSON.stringify({ strict: true })),
    ).rejects.toBeTruthy();
  });
});
